zed: Support opening welcome page on startup

tidely opened the pull request:
Closes #40100 #43158

I saw a great suggestion in #43158 of reusing the welcome page as an option when opening Zed. Preferably we would have  a dedicated startup page at some later point in time that could highlight other useful features such as remoting (or add it to the welcome page?).

Currently still on the todo list:

- [ ] Think about removing `Return to Setup` from welcome screen conditionally (feedback appreciated)
- [ ] Consider whether Welcome Page should be the default, a setting, used as a fallback or something else entirely, it likely should not be under "what to restore from a previous session"
- [ ] Docs

The PR will likely be decently large, I'm willing to pair with someone to get it merged

<img width="1509" height="842" alt="image" src="https://github.com/user-attachments/assets/5ca204da-675c-4e39-adad-08bfca07ee0f" />
<img width="1296" height="372" alt="image" src="https://github.com/user-attachments/assets/88a6f86b-8551-4867-b210-20d613075b28" />

Release Notes:

- New setting to open a welcome page when launching Zed

tidely committed:
support for opening welcome page on launch

tidely committed:
unused import

tidely committed:
undo

VihaanReddyM commented:
Yes please add this feature it will give a nice vibe. I tried adding this but was a little hard for me because I don't really use rust

VihaanReddyM commented:
I’ve thought about it a bit more and here’s what I suggest:

* The **“Return to Setup”** button should be hidden when the welcome screen is shown after the initial setup flow. It only makes sense the first time.
* For the behavior of when to show the welcome screen, we could add a **new setting** that lets users choose between two modes:

  1. Show the welcome screen only when no workspace is open (current behavior).
  2. Always show the welcome screen on startup.

This setting should be exposed both in the settings file and the UI menu.
I tried implementing this logic earlier, but I got stuck on wiring it into the settings system.

Let me know if this direction makes sense.

tidely commented:
> I have just built your fork and there is a issue, When zed is first launched it is skipping the setup page completely so gotta work on that.

Unless you deleted your main Zed installation, I think it will detect that it's not your first time installing Zed

tidely commented:
> * For the behavior of when to show the welcome screen, we could add a **new setting** that lets users choose between two modes:
>   
>   1. Show the welcome screen only when no workspace is open (current behavior).
>   2. Always show the welcome screen on startup.
> 
> This setting should be exposed both in the settings file and the UI menu. I tried implementing this logic earlier, but I got stuck on wiring it into the settings system.

I think the most intuitive solution would be to allow putting all startup options into any order you want. However that seems like something maybe not possible with the current settings UI?

tidely committed:
remove onboarding after adding welcome page to

onboarding is stored as a entityid which allows us to find it regardless of the index, however this is not true for the welcome page, meaning it would previous go out of order if the onboarding index was smaller than the welcome page index

tidely committed:
deduplicate restorable_workspace_locations logic

tidely committed:
clean error logic

VihaanReddyM commented:
> > * For the behavior of when to show the welcome screen, we could add a **new setting** that lets users choose between two modes:
> 
> >   
> 
> >   1. Show the welcome screen only when no workspace is open (current behavior).
> 
> >   2. Always show the welcome screen on startup.
> 
> > 
> 
> > This setting should be exposed both in the settings file and the UI menu. I tried implementing this logic earlier, but I got stuck on wiring it into the settings system.
> 
> 
> 
> I think the most intuitive solution would be to allow putting all startup options into any order you want. However that seems like something maybe not possible with the current settings UI?
> 
> 
> 
> 

Idk what I did was I added a dropdown.

franciskafyi commented:
Thank you so much @tidely. We agree this is needed but we actually prefer this other approach in here: https://github.com/zed-industries/zed/pull/44048. Here, the page is adapted slightly with a focus on Recent Projects. That makes more sense for Existing Users, I think.
